keps/sig-architecture/2558-field-version-openapi/README.md

```diff
@@ -0,0 +1,414 @@
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a way for publishing field-level versioning
+information in OpenAPI using comment tags on API fields.
+
+Specifically, the field-level versioning information would include the
+prerelease status, minimum Kubernetes version for the prerelease status
+and the feature gate associated with field.
+
+## Motivation
+
+**Feature gate**
+
+When a new field is added to an existing API version, a feature gate
+is added to control the enablement of the new field.
+
+The feature gates are defined in [`kube_features.go`]. The comments
+for each of these feature gates in [`kube_features.go`] lists
+the prerelease status of the feature gate and the minimum Kubernetes
+version for the particular prerelease. These comments do not follow
+a pattern and are not parseable.
+
+Currently, when a new field is added, there is no way to map and document
+the field with the feature gate that controls it. Some fields mention
+the feature gates in a comment but these comments are not parseable either.
+
+**Field version**
+
+Currently, when a new field is added to a stable API version, there
+is no way to expose the version of the new field in OpenAPI.
+One way to identify alpha API fields is by comparing the OpenAPI schema
+generated from an apiserver with all APIs enabled and the one generated
+from an apiserver with only stable and beta APIs enabled.
+However, this process is cumbersome.
```

Comment:
and it is not always feasible ... not all alpha/beta fields are gated at API group level.